client/src/App.js

```diff
@@ -1,12 +1,210 @@
+          'Last 24 Hours': now - 86400000,
+          'Last Week': now - (7 * 86400000),
+          'Last Month': now - (31 * 86400000),
+          'Last Year': now - (345 * 86400000),
+          'Last 2 Years': now - (345 * 86400000 * 2),
+          'Last 5 Years': now - (345 * 86400000 * 5),
+          Always: 0,
+        };
+        return (creationTime > timeRangeValues[timeRange.range]);
+      };
+      filteredTickets = filteredTickets.filter(filterByTime);
+    }
+    if (filterLabels.some((label) => label.active)) { // filter tickets by chosen labels
+      const activeLabels = filterLabels.filter((label) => label.active);
+      const filterByLabels = (ticket) => {
+        if (!ticket.labels) return false;
+        return activeLabels.every((label) => ticket.labels.includes(label.name));
+      };
+      filteredTickets = filteredTickets.filter(filterByLabels);
+    }
+
+    return filteredTickets.map((ticket) => (
+      <Ticket
+        data={ticket}
+        onHide={hideTicket}
+        update={grabTickets}
+        options={options}
+        labelClick={labelClick}
+      />
+    ));
+  }
+  function unHideTickets() {
+    const newTickets = ticketsToDisplay.map((ticket) => {
+      if (ticket.hide) { ticket.hide = false; }
+      return ticket;
+    });
+    setTicketsToDisplay(newTickets);
+  }
+  async function searchTickets(e) {
+    const query = e.target.value;
+    const tickets = (await axios.get(`/api/tickets?searchText=${query}`)).data;
+    setTicketsToDisplay(tickets);
+    getAvailableLabels(tickets);
+  }
+  const activeLabelFilters = options.filterLabels.filter((label) => label.active);
+  const ticketsHidden = ticketsToDisplay.filter((ticket) => ticket.hide);
+
   return (
-    <main>
-    </main>
+    <>
+      <header style={{ background: options.ThemeColor.color }}>
+
+        <img id="logo" src={icon} alt="icon" />
+        <h1> Ticket Master</h1>
+        <Search
+          search={searchTickets}
+          id="searchInput"
+          placeholder="search a title"
+        />
+      </header>
+      <div id="optionsStatus">
+        <span id="labelStatus" className="status">
+          filter by labels:
+          {
+            activeLabelFilters.length
+              ? (
+                <ul>
+                  {
+                    activeLabelFilters
+                      .map((label) => <li key={label.name}>{label.name}</li>)
+                  }
+                </ul>
+              )
+              : 'none'
+          }
+        </span>
+        <span id="doneStatus" className="status">
+          closed tickets:
+          {options.hideDone.active ? 'hidden' : 'shown'}
+        </span>
+        <span id="timeStatus" className="status">
+          time range:
+          {options.timeRange.active ? options.timeRange.range : 'all'}
+        </span>
+      </div>
+
+      <div id="ticketCounterContainer">
+        <span id="ticketCounter">
+          {renderTickets().length}
```

Comment:
really bad, you render all just for getting the length?  try to call renderTickets once + I know this how react act in their tutorial but I can't see why to move out this logic to an external

Reply:
it is done so i always have the updated count displayed in the counter.
I tried making a variable ticketCount which the renderTickets function sets when it is called, but since the counter is rendered before the tickets are ( before the function is called)  it means it isn't set when I need it. it can be fixed with moving the counter down in the code and then back up with some css magic...
do you think it would be better to make another function which does basically the same as renderTickets but doesn't render in the end in order to get the ticket count?

Reply:
i changed it so something else.